Gems/ROS2/Code/Source/Frame/ROS2FrameSystemComponent.cpp

```diff
@@ -0,0 +1,440 @@
+        AZStd::set<AZ::EntityId>& frameToUnregisterWatchedEntities = m_watchedEntities.find(frameToUnregister)->second;
+        const AZ::EntityId& frameToUnregisterParent = m_frameParent.find(frameToUnregister)->second;
+        AZStd::set<AZ::EntityId>& parentsChildren = m_frameChildren.find(frameToUnregisterParent)->second;
+
+        if (frameToUnregisterChildren.empty())
+        {
+            for (const AZ::EntityId& watchedEntity : frameToUnregisterWatchedEntities)
+            {
+                ROS2FrameSystemTransformHandler& handler = m_watchedEntitiesHandlers.find(watchedEntity)->second;
+                handler.RemoveFrameEntity(frameToUnregister);
+                if (handler.GetFrameAmount() == 0)
+                {
+                    handler.BusDisconnect();
+                    m_watchedEntitiesHandlers.erase(watchedEntity);
+                }
+            }
+        }
+        else
+        {
+            for (const AZ::EntityId child : frameToUnregisterChildren)
+            {
+                // Change the handlers and watched entities
+                AZStd::set<AZ::EntityId>& childsWatchedEntities = m_watchedEntities.find(child)->second;
+                for (const AZ::EntityId& watchedEntity : frameToUnregisterWatchedEntities)
+                {
+                    childsWatchedEntities.insert(watchedEntity);
+                    ROS2FrameSystemTransformHandler& handler = m_watchedEntitiesHandlers.find(watchedEntity)->second;
+                    handler.RemoveFrameEntity(frameToUnregister);
+                    handler.AddFrameEntity(child);
+                }
+
+                // Change the parenthood
+                parentsChildren.insert(child);
+                m_frameParent.find(child)->second = frameToUnregisterParent;
+            }
+        }
+
+        // Update namespaces before removal of children
+        AZStd::string ns;
+        ROS2FrameComponentBus::EventResult(ns, frameToUnregisterParent, &ROS2FrameComponentBus::Events::GetNamespace);
+        UpdateNamespaces(frameToUnregister, ns, false);
+
+        // Remove all references to the unregisteredEntity
+        m_frameChildren.erase(frameToUnregister);
+        m_frameParent.erase(frameToUnregister);
+        m_watchedEntities.erase(frameToUnregister);
+
+        // Remove itself form the parents children
+        parentsChildren.erase(frameToUnregister);
+
+        // Check if the parents children are empty to ensure that the level parent will be deleted
+        if (parentsChildren.size() == 0)
+        {
+            if (m_frameParent.find(frameToUnregisterParent)->second == frameToUnregisterParent)
+            { // Remove the level entity from the registry
+                m_frameChildren.erase(frameToUnregisterParent);
+                m_frameParent.erase(frameToUnregisterParent);
+            }
+        }
+    }
+
+    void ROS2FrameSystemComponent::MoveFrame(const AZ::EntityId& frameEntityId, const AZ::EntityId& newParent)
```

Comment:
I would consider splitting body of this method into two helper private methods - remove and add. This may improve overall readability.